pkg/dockerregistry/client.go

```diff
@@ -226,8 +242,112 @@ func (c *connection) getCachedRepository(name string) (*repository, error) {
 	return repo, nil
 }
 
-func (c *connection) getRepository(name string) (*repository, error) {
+func (c *connection) checkV2() (bool, error) {
+	base := c.url
+	base.Path = path.Join(base.Path, "v2")
+	req, err := http.NewRequest("GET", base.String(), nil)
+	if err != nil {
+		return false, fmt.Errorf("error creating request: %v", err)
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		// if we tried https and were rejected, try http
+		if c.url.Scheme == "https" && c.allowInsecure {
+			glog.V(4).Infof("Failed to get https, trying http: %v", err)
+			c.url.Scheme = "http"
+			return c.checkV2()
+		}
+		return false, convertConnectionError(c.url.String(), fmt.Errorf("error checking for V2 registry at %s: %v", base.String(), err))
+	}
+	defer resp.Body.Close()
+
+	switch code := resp.StatusCode; {
+	case code == http.StatusUnauthorized:
+		// handle auth challenges on individual repositories
+	case code == http.StatusNotFound:
+		fallthrough
+	case code >= 300 || resp.StatusCode < 200:
+		return false, nil
+	}
+	// TODO: check Docker-Distribution-API-Version?
+	return true, nil
+}
+
+func parseAuthChallenge(header string) (map[string]string, bool) {
+	header = strings.ToLower(header)
+	if !strings.HasPrefix(header, "bearer ") {
```

Comment:
Too complex, this isn't to handle any auth except simple token for
dockerhub.  But I'll return a clear error nowZ

On Aug 11, 2015, at 11:26 AM, Andy Goldstein <notifications@github.com>
wrote:

In pkg/dockerregistry/client.go
<https://github.com/openshift/origin/pull/4104#discussion_r36758303>:

> +
> +	switch code := resp.StatusCode; {
> +	case code == http.StatusUnauthorized:
> +		// handle auth challenges on individual repositories
> +	case code == http.StatusNotFound:
> +		fallthrough
> +	case code >= 300 || resp.StatusCode < 200:
> +		return false, nil
> +	}
> +	// TODO: check Docker-Distribution-API-Version?
> +	return true, nil
> +}
> +
> +func parseAuthChallenge(header string) (map[string]string, bool) {
> +	header = strings.ToLower(header)
> +	if !strings.HasPrefix(header, "bearer ") {

The daemon supports both basic and bearer token auth for v2

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4104/files#r36758303>.